pkg/internal/cnsoperator/cnsfilevolumeclient/cnsfilevolumeclient.go

```diff
@@ -0,0 +1,246 @@
+// the list of clientVMNames that expose the same external IP
+// address for a given CnsFileVolumeClient instance.
+// Callers need to specify fileVolumeName as a combination of
+// "<SV-namespace>/<SV-PVC-name>". This combination is used to uniquely
+// identify CnsFileVolumeClient instances.
+// If the given VM was the last client for this file volume, the instance is deleted from
+// the API server.
+// Returns an error if the operation cannot be persisted on the API server.
+func (f *fileVolumeClient) RemoveClientVMFromIPList(ctx context.Context, fileVolumeName, clientVMName, clientVMIP string) error {
+	log := logger.GetLogger(ctx)
+	log.Infof("Removing clientVM %s from cnsfilevolumeclient %s list for IP address %s", clientVMName, fileVolumeName, clientVMIP)
+	actual, _ := f.volumeLock.LoadOrStore(fileVolumeName, &sync.Mutex{})
+	instanceLock, ok := actual.(*sync.Mutex)
+	if !ok {
+		return fmt.Errorf("failed to cast lock for cnsfilevolumeclient instance: %s", fileVolumeName)
+	}
+	instanceLock.Lock()
+	defer instanceLock.Unlock()
+	obj, exists, err := f.store.GetByKey(fileVolumeName)
+	if err != nil {
+		log.Errorf("failed to get key %s from store with error: %+v", fileVolumeName, err)
+		return err
+	}
+	if !exists {
+		log.Infof("cnsfilevolumeclient instance %s does not exist on API server", fileVolumeName)
+		return nil
+	}
+	instance := &v1alpha1.CnsFileVolumeClient{}
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(*unstructured.Unstructured).UnstructuredContent(), instance)
+	if err != nil {
+		log.Errorf("failed to cast to cnsfilevolumeclient from unstructured")
+		return err
+	}
+	log.Debugf("Verifying if clientVM %s exists in ExternalIPtoClientVms list for IP address: %s. Current list: %+v", clientVMName, clientVMIP, instance.Spec.ExternalIPtoClientVms[clientVMIP])
+	for index, existingClientVM := range instance.Spec.ExternalIPtoClientVms[clientVMIP] {
+		if clientVMName == existingClientVM {
+			log.Debugf("Removing clientVM %s from ExternalIPtoClientVms list", clientVMName)
+			instance.Spec.ExternalIPtoClientVms[clientVMIP] = append(instance.Spec.ExternalIPtoClientVms[clientVMIP][:index], instance.Spec.ExternalIPtoClientVms[clientVMIP][index+1:]...)
+			if len(instance.Spec.ExternalIPtoClientVms[clientVMIP]) == 0 {
+				log.Debugf("Deleting entry for IP %s from spec.ExternalIPtoClientVms", clientVMIP)
+				delete(instance.Spec.ExternalIPtoClientVms, clientVMIP)
+			}
+			if len(instance.Spec.ExternalIPtoClientVms) == 0 {
+				log.Debugf("Deleting cnsfilevolumeclient instance %s from API server", fileVolumeName)
+				err = f.client.Delete(ctx, instance)
+				if err != nil {
+					log.Errorf("failed to delete cnsfilevolumeclient instance %s with error: %+v", fileVolumeName, err)
+					return err
+				}
+				f.volumeLock.Delete(fileVolumeName)
+				return nil
+			}
+			log.Debugf("Updating cnsfilevolumeclient instance %s with spec: %+v", fileVolumeName, instance)
+			err = f.client.Update(ctx, instance)
```

Comment:
Same comment here as well. We are updating the CR in the API server but `AddClientVMToIPList` is taking decisions based on what is in the cache store which could be stale.